### PR TITLE
Support user info in Elasticsearch URIs

### PR DESCRIPTION
**Pull request by evgeniycheban**
Fixes gh-21291

### CODE REVIEW

**Comment by wilkinsona** [comment]
Thanks very much for the PR, @evgeniycheban 

Could you please take a look at adding some tests for the new functionality? I think they could be implemented as additions to `org.springframework.boot.autoconfigure.elasticsearch.ElasticsearchRestClientAutoConfigurationTests`. It looks like you can get the nodes from the `RestClient` to verify that the configuration has been applied correctly. As part of the tests, it would be good to verify the behaviour of URIs with and without any user info and how the separate username and password properties are used.

**Comment by evgeniycheban** [comment]
> Thanks very much for the PR, @evgeniycheban
> 
> Could you please take a look at adding some tests for the new functionality? I think they could be implemented as additions to `org.springframework.boot.autoconfigure.elasticsearch.ElasticsearchRestClientAutoConfigurationTests`. It looks like you can get the nodes from the `RestClient` to verify that the configuration has been applied correctly. As part of the tests, it would be good to verify the behaviour of URIs with and without any user info and how the separate username and password properties are used.

Thanks for the review, @wilkinsona 

Sure, I will add tests.

**Comment by evgeniycheban** [comment]
@wilkinsona I added tests.

When you have a moment, please take a look.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java [hunk 1]
```diff
@@ -114,25 +123,48 @@ RestClient elasticsearchRestClient(RestClientBuilder builder) {
 
 		private final ElasticsearchRestClientProperties properties;
 
+		private CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java:126]: Creating a field for the `CredentialsProvider` and allowing it to be injected doesn't seem to be related to supported user info in the URI. Can you please revert this part of the changes?

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java [hunk 2]
```diff
@@ -156,6 +163,69 @@ void restClientCanQueryElasticsearchNode() {
...
+
+					CredentialsProvider credentialsProvider = context.getBean(CredentialsProvider.class);
+
+					Credentials uriCredentials = credentialsProvider.getCredentials(new AuthScope("localhost", 9200));
+					assertThat(uriCredentials.getUserPrincipal().getName()).isEqualTo("user");
+					assertThat(uriCredentials.getPassword()).isEqualTo("password");
+
+					Credentials defaultCredentials = credentialsProvider
+							.getCredentials(new AuthScope("localhost", 9201));
+					assertThat(defaultCredentials.getUserPrincipal().getName()).isEqualTo("admin");
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java:214]: I think the username in the URI should win so the name of the principal should be `user`.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java [hunk 3]
```diff
@@ -156,6 +163,69 @@ void restClientCanQueryElasticsearchNode() {
...
+					assertThat(defaultCredentials.getPassword()).isEqualTo("admin");
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java:215]: Similarly, I think the password in the URI should win so it should be `password`.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java (hunk 2 above)
Comment (evgeniycheban) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java:214]: Yes, that is right, in this assert we are checking that if the URI does not contain credentials then `spring.elasticsearch.rest.username` and `spring.elasticsearch.rest.password` will be used.

Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java:214]: Sorry, I'd missed that there are two URIs being configured, one with credentials and one without.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java (hunk 3 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientAutoConfigurationTests.java:215]: Sorry, ignore this. I'd missed that there are two URIs being configured, one with credentials and one without.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java (hunk 1 above)
Comment (evgeniycheban) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java:126]: I added this to make testing easer, because we can not access `credentialsProvider` field from `InternalHttpAsyncClient`. It can be also useful in some cases when the user needs to use a custom `CredentialsProvider`.

Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java:126]: I don't think it's safe to expose a `CredentialProvider` as a bean. In this case the credentials that it provides are specific to Elasticsearch, but there's nothing about the bean that indicates that's the case. It could be accidentally consumed and used anywhere that's working with an Apache HTTP client, potentially leaking credentials to another service.

For this sort of situation, we prefer to use AssertJ's support for extracting fields. In this case it would look something like this:

```java
assertThat(client).extracting("client")
		.extracting("credentialsProvider", InstanceOfAssertFactories.type(CredentialsProvider.class))
		.satisfies((credentialsProvider) -> {
			Credentials uriCredentials = credentialsProvider.getCredentials(new AuthScope("localhost", 9200));
			assertThat(uriCredentials.getUserPrincipal().getName()).isEqualTo("user");
			assertThat(uriCredentials.getPassword()).isEqualTo("password");
			Credentials defaultCredentials = credentialsProvider.getCredentials(new AuthScope("localhost", 9201));
			assertThat(defaultCredentials.getUserPrincipal().getName()).isEqualTo("admin");
			assertThat(defaultCredentials.getPassword()).isEqualTo("admin");
		});
```

Comment (evgeniycheban) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java:126]: Thank you for the AssertJ example, @wilkinsona 

I will update tests.

Comment (evgeniycheban) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/ElasticsearchRestClientConfigurations.java:126]: Fixed.

**Comment by evgeniycheban** [comment]
@wilkinsona Could you please take a look at the latest changes when you have a moment?

**Comment by murdos** [comment]
@wilkinsona : is there any chance to get this in 2.3.x ? 
Indeed this feature was available with Jest, but support has been removed in 2.3.

**Comment by wilkinsona** [comment]
Yes. Given the situation with Jest, it seems reasonable to include this in 2.3.x.

**Comment by philwebb** [comment]
Thanks very much for the PR @evgeniycheban, this is now in 2.3.x and master. I made some minor tweaks in f8982bdc14af90b590c8131c09b95b05edebdd79 (mainly to protect against URL parse exceptions).
